Sources/SwiftKafka/KafkaConsumer.swift

```diff
@@ -0,0 +1,418 @@
+    private let messagesSource: NIOAsyncSequenceProducer<
+        Element,
+        ConsumerMessagesAsyncSequence.HighLowWatermark,
+        ConsumerMessagesAsyncSequenceDelegate
+    >.Source
+    /// `AsyncSequence` that returns all ``KafkaConsumerMessage`` objects that the consumer receives.
+    public nonisolated let messages: ConsumerMessagesAsyncSequence
+
+    /// Initialize a new ``KafkaConsumer``.
+    /// To listen to incoming messages, please subscribe to a list of topics using ``subscribe(topics:)``
+    /// or assign the consumer to a particular topic + partition pair using ``assign(topic:partition:offset:)``.
+    /// - Parameter config: The ``KafkaConfig`` for configuring the ``KafkaConsumer``.
+    /// - Parameter logger: A logger.
+    private init(
+        config: KafkaConfig = KafkaConfig(),
+        logger: Logger
+    ) async throws {
+        self.config = config
+        self.logger = logger
+        self.client = try KafkaClient(type: .consumer, config: self.config, logger: self.logger)
+
+        self.subscribedTopicsPointer = rd_kafka_topic_partition_list_new(1)
+
+        // Events that would be triggered by rd_kafka_poll
+        // will now be also triggered by rd_kafka_consumer_poll
+        let result = self.client.withKafkaHandlePointer { handle in
+            rd_kafka_poll_set_consumer(handle)
+        }
+        guard result == RD_KAFKA_RESP_ERR_NO_ERROR else {
+            throw KafkaError(rawValue: result.rawValue)
+        }
+
+        let backpressureStrategy = ConsumerMessagesAsyncSequence.HighLowWatermark(
+            lowWatermark: 5,
+            highWatermark: 10
+        )
+
+        self.state = .started
+
+        let messagesSequenceDelegate = ConsumerMessagesAsyncSequenceDelegate()
+        let messagesSourceAndSequence = NIOAsyncSequenceProducer.makeSequence(
+            of: Element.self,
+            backPressureStrategy: backpressureStrategy,
+            delegate: messagesSequenceDelegate
+        )
+        self.messagesSource = messagesSourceAndSequence.source
+        self.messages = ConsumerMessagesAsyncSequence(
+            wrappedSequence: messagesSourceAndSequence.sequence
+        )
+
+        await messagesSequenceDelegate.setConsumer(self)
+    }
+
+    /// Initialize a new ``KafkaConsumer`` and subscribe to the given list of `topics` as part of
+    /// the consumer group specified in `groupID`.
+    /// - Parameter topics: An array of topic names to subscribe to.
+    /// - Parameter groupID: Name of the consumer group that this ``KafkaConsumer`` will create / join.
+    /// - Parameter config: The ``KafkaConfig`` for configuring the ``KafkaConsumer``.
+    /// - Parameter logger: A logger.
+    public convenience init(
+        topics: [String],
+        groupID: String,
+        config: KafkaConfig = KafkaConfig(),
+        logger: Logger
+    ) async throws {
+        // TODO: make prettier
+        // TODO: use inout parameter here?
+        var config = config
```

Comment:
I am not really happy with how I am modifying the `config` here. What do you think?